src/tests/dotnet-counters/CounterMonitorPayloadTests.cs

```diff
@@ -0,0 +1,311 @@
+                trace.events.Where(e => e.name.Equals(Constants.TestHistogramName)).Select(e => e.tags).Where(t => !string.IsNullOrEmpty(t)).ToHashSet(),
+                trace.events.Where(e => e.name.Equals(Constants.TestCounterName)).Select(e => e.tags).Where(t => !string.IsNullOrEmpty(t)).ToHashSet(),
+                trace.events.Where(e => e.name.Equals(Constants.TestHistogramName)).Select(e => e.value).ToList(),
+                trace.events.Where(e => e.name.Equals(Constants.TestCounterName)).Select(e => e.value).ToList(),
+                CountersExportFormat.json
+                );
+        }
+
+        [SkippableTheory, MemberData(nameof(Configurations))]
+        public async Task TestCounterMonitorCustomMetricsCSV(TestConfiguration configuration)
+        {
+            CheckFramework(configuration.BuildProjectFramework);
+
+            List<string> lines = await GetCounterTraceCSV(configuration, new List<string> { Constants.TestMeterName });
+            IEnumerable<string[]> splitLines = lines.Select(l => l.Split(","));
+
+            var providers = splitLines.Select(line => line[Constants.ProviderIndex]).ToHashSet();
+
+            var countersList = splitLines.Select(line => line[Constants.CounterNameIndex]).ToList();
+            var counterNames = countersList.Select(counter => counter.Split(TagStart)[0]).ToHashSet();
+
+            var counterTypes = splitLines.Select(line => line[Constants.CounterTypeIndex]).ToHashSet();
+
+            var counterTags = GetCSVTags(countersList, Constants.TestCounterName);
+            var histogramTags = GetCSVTags(countersList, Constants.TestHistogramName);
+
+            var counterValues = GetCSVValues(splitLines, Constants.TestCounterName);
+            var histogramValues = GetCSVValues(splitLines, Constants.TestHistogramName);
+
+            ValidateCustomMetrics(
+                providers,
+                counterNames,
+                counterTypes,
+                histogramTags,
+                counterTags,
+                histogramValues,
+                counterValues,
+                CountersExportFormat.csv);
+        }
+
+        [SkippableTheory, MemberData(nameof(Configurations))]
+        public async Task TestCounterMonitorSystemRuntimeMetricsJSON(TestConfiguration configuration)
+        {
+            JSONCounterTrace trace = await GetCounterTraceJSON(configuration, new List<string> { SystemRuntimeName });
+            Assert.NotEmpty(trace.events);
+
+            ValidateSystemRuntimeMetrics(trace.events.Select(e => e.provider).Distinct().ToHashSet(), trace.events.Select(e => e.name).Distinct().ToHashSet(), trace.events.Select(e => e.counterType).Distinct());
+        }
+
+        [SkippableTheory, MemberData(nameof(Configurations))]
+        public async Task TestCounterMonitorSystemRuntimeMetricsCSV(TestConfiguration configuration)
+        {
+            List<string> lines = await GetCounterTraceCSV(configuration, new List<string> { SystemRuntimeName });
```

Comment:
Consider doing at least some of the parsing inside GetCounterTraceCSV and returning an object model rather than just an array of lines.